ledger/src/blockstore.rs

```diff
@@ -3598,6 +3598,84 @@ impl Blockstore {
         self.get_slot_entries_in_block(slot, vec![(start_index, end_index)], slot_meta)
     }
 
+    /// Returns the merkle roots from the last `DATA_SHREDS_PER_FEC_BLOCK`
+    /// data shreds of a slot.
+    /// Will fail if:
+    ///     - Slot meta is missing
+    ///     - LAST_SHRED_IN_SLOT flag has not been received
+    ///     - There are missing shreds in the last fec set
+    /// If there are fewer than `DATA_SHREDS_PER_FEC_BLOCK` shreds in the slot
+    /// this will return all that are available.
+    fn get_last_merkle_roots(&self, slot: Slot) -> Result<Vec<Option<Hash>>> {
+        let slot_meta = self
+            .meta_cf
+            .get(slot)?
+            .ok_or(BlockstoreError::SlotUnavailable)?;
+        let last_shred_index = slot_meta
+            .last_index
+            .ok_or(BlockstoreError::InvalidShredData(Box::new(
+                bincode::ErrorKind::Custom(format!(
+                    "last shred index is missing for {slot} {slot_meta:?}"
+                )),
+            )))?;
+        let num_shreds = u64::try_from(DATA_SHREDS_PER_FEC_BLOCK)
+            .expect("DATA_SHREDS_PER_FEC_BLOCK should fit in u64");
+        let start_index = last_shred_index
+            .saturating_sub(num_shreds)
+            .saturating_add(1);
+        let keys: Vec<(Slot, u64)> = (start_index..=last_shred_index)
+            .map(|index| (slot, index))
+            .collect();
+
+        self.data_shred_cf
+            .multi_get_bytes(keys)
+            .into_iter()
+            .enumerate()
+            .map(|(idx, shred_bytes)| {
+                let shred_bytes = shred_bytes.ok().flatten();
+                if shred_bytes.is_none() {
+                    return Err(BlockstoreError::InvalidShredData(Box::new(
+                        bincode::ErrorKind::Custom(format!(
+                            "Missing shred for slot {slot}, index {idx}"
+                        )),
+                    )));
+                }
+                Ok(shred::layout::get_merkle_root(&shred_bytes.unwrap()))
```

Comment:
The `.unwrap` can be avoided with `.ok_or_else(...)?`.
also, if we convert `None` to `Err` this can just return a `Vec<Result<Hash>>` and short-circuit as soon as `get_merkle_root` is `None`.